File: python/ingest_benchmark.py
#!/usr/bin/env python3
"""
Reference ingestion benchmark pipeline for cross-language comparison.

Pipeline:
JSONL(.gz) -> validate -> transform -> aggregate -> Parquet + checksum
"""

from __future__ import annotations

import glob
import gzip
import hashlib
import json
import os
from dataclasses import dataclass
from datetime import datetime, timedelta
from pathlib import Path
from typing import Dict, Iterable, List, Tuple
from urllib.request import Request, urlopen

import pyarrow as pa
import pyarrow.parquet as pq

# Edit these values directly instead of passing CLI arguments.
CONFIG = {
    "MODE": "cpu",  # "cpu" (strict validation) or "io" (relaxed validation)
    "INPUT_GLOBS": [],  # Example: ["data/raw/gharchive/*.json.gz"]
    "DOWNLOAD_GHARCHIVE_DATE": "2026-01-01",  # Set to None to disable downloads
    "DOWNLOAD_HOURS": 10,
    "DOWNLOAD_DIR": "data/raw/gharchive",
    "OUTPUT_PARQUET": "data/out/python_cpu_1h.parquet",
    "OUTPUT_CHECKSUM": "data/out/python_cpu_1h.sha256",
}


AggregateKey = Tuple[str, str]
AggregateValue = Tuple[int, int]


@dataclass
class Stats:
    total_lines: int = 0
    valid_lines: int = 0
    invalid_lines: int = 0


def ensure_parent(path: str) -> None:
    Path(path).parent.mkdir(parents=True, exist_ok=True)


def download_gharchive(date_str: str, hours: int, download_dir: str) -> List[str]:
    if hours <= 0:
        return []
    date_obj = datetime.strptime(date_str, "%Y-%m-%d")
    os.makedirs(download_dir, exist_ok=True)
    downloaded = []
    for hour in range(hours):
        ts = date_obj + timedelta(hours=hour)
        url = f"https://data.gharchive.org/{ts:%Y-%m-%d}-{ts.hour}.json.gz"
        out_path = os.path.join(download_dir, f"{ts:%Y-%m-%d}-{ts.hour}.json.gz")
        if not os.path.exists(out_path):
            print(f"[download] {url} -> {out_path}")
            req = Request(url, headers={"User-Agent": "Mozilla/5.0"})
            with urlopen(req) as resp, open(out_path, "wb") as out_f:
                out_f.write(resp.read())
        else:
            print(f"[download] skip existing {out_path}")
        downloaded.append(out_path)
    return downloaded


def resolve_input_files(patterns: Iterable[str]) -> List[str]:
    files: List[str] = []
    for pattern in patterns:
        matches = sorted(glob.glob(pattern))
        files.extend(matches)
    return sorted(set(files))


def validate_record(record: dict, strict: bool) -> bool:
    if not isinstance(record, dict):
        return False

    record_id = record.get("id")
    event_type = record.get("type")
    created_at = record.get("created_at")
    repo = record.get("repo")

    if not isinstance(record_id, str) or not record_id:
        return False
    if not isinstance(event_type, str) or not event_type:
        return False
    if not isinstance(created_at, str) or not created_at:
        return False
    if not isinstance(repo, dict):
        return False
    if "id" not in repo:
        return False

    repo_id = repo.get("id")
    try:
        int(repo_id)
    except (TypeError, ValueError):
        return False

    if strict:
        # GH Archive timestamps are expected as ISO-like UTC strings.
        if "T" not in created_at or not created_at.endswith("Z"):
            return False
        if len(event_type) > 100:
            return False
    return True


def transform_record(record: dict) -> Tuple[str, str, int]:
    created_at = record["created_at"]
    event_day = created_at[:10]
    event_type = str(record["type"]).strip().lower()
    repo_id = int(record["repo"]["id"])
    return event_day, event_type, repo_id


def process_files(input_files: List[str], strict: bool) -> Tuple[Dict[AggregateKey, AggregateValue], Stats]:
    aggregates: Dict[AggregateKey, AggregateValue] = {}
    stats = Stats()

    for file_path in input_files:
        print(f"[process] {file_path}")
        with gzip.open(file_path, mode="rt", encoding="utf-8") as f:
            for line in f:
                stats.total_lines += 1
                line = line.strip()
                if not line:
                    stats.invalid_lines += 1
                    continue
                try:
                    record = json.loads(line)
                except json.JSONDecodeError:
                    stats.invalid_lines += 1
                    continue

                if not validate_record(record, strict=strict):
                    stats.invalid_lines += 1
                    continue

                stats.valid_lines += 1
                event_day, event_type, repo_id = transform_record(record)
                key = (event_day, event_type)
                count, repo_sum = aggregates.get(key, (0, 0))
                aggregates[key] = (count + 1, repo_sum + repo_id)

    return aggregates, stats


def write_parquet(output_parquet: str, rows: List[Tuple[str, str, int, int]]) -> None:
    ensure_parent(output_parquet)
    table = pa.table(
        {
            "event_day": [r[0] for r in rows],
            "event_type": [r[1] for r in rows],
            "record_count": [r[2] for r in rows],
            "sum_repo_id": [r[3] for r in rows],
        }
    )
    pq.write_table(table, output_parquet, compression="zstd")


def write_checksum(output_checksum: str, rows: List[Tuple[str, str, int, int]]) -> str:
    ensure_parent(output_checksum)
    digest = hashlib.sha256()
    for row in rows:
        digest.update(f"{row[0]}|{row[1]}|{row[2]}|{row[3]}\n".encode("utf-8"))
    checksum = digest.hexdigest()
    with open(output_checksum, "w", encoding="utf-8") as f:
        f.write(checksum + "\n")
    return checksum


def main() -> None:
    mode = str(CONFIG["MODE"]).lower()
    if mode not in {"cpu", "io"}:
        raise SystemExit("CONFIG['MODE'] must be either 'cpu' or 'io'.")
    strict = mode == "cpu"

    downloaded: List[str] = []
    download_date = CONFIG["DOWNLOAD_GHARCHIVE_DATE"]
    download_hours = int(CONFIG["DOWNLOAD_HOURS"])
    download_dir = str(CONFIG["DOWNLOAD_DIR"])
    if download_date and download_hours > 0:
        downloaded = download_gharchive(str(download_date), download_hours, download_dir)

    input_patterns = [str(p) for p in CONFIG["INPUT_GLOBS"]]
    input_files = resolve_input_files(input_patterns) + downloaded
    input_files = sorted(set(input_files))
    if not input_files:
        raise SystemExit("No input files found. Set CONFIG['INPUT_GLOBS'] and/or download settings.")

    aggregates, stats = process_files(input_files=input_files, strict=strict)
    sorted_rows = sorted((k[0], k[1], v[0], v[1]) for k, v in aggregates.items())

    output_parquet = str(CONFIG["OUTPUT_PARQUET"])
    output_checksum = str(CONFIG["OUTPUT_CHECKSUM"])
    write_parquet(output_parquet, sorted_rows)
    checksum = write_checksum(output_checksum, sorted_rows)

    print("[done]")
    print(f"mode={mode}")
    print(f"input_files={len(input_files)}")
    print(f"total_lines={stats.total_lines}")
    print(f"valid_lines={stats.valid_lines}")
    print(f"invalid_lines={stats.invalid_lines}")
    print(f"aggregate_rows={len(sorted_rows)}")
    print(f"checksum={checksum}")
    print(f"parquet={output_parquet}")
    print(f"checksum_file={output_checksum}")


if __name__ == "__main__":
    main()

File: java/src/main/java/org/tud/IngestBenchmark.java
package org.tud;

import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;
import org.apache.avro.Schema;
import org.apache.avro.generic.GenericData;
import org.apache.avro.generic.GenericRecord;
import org.apache.hadoop.conf.Configuration;
import org.apache.hadoop.fs.Path;
import org.apache.parquet.avro.AvroParquetWriter;
import org.apache.parquet.hadoop.ParquetWriter;
import org.apache.parquet.hadoop.metadata.CompressionCodecName;

import java.io.BufferedReader;
import java.io.BufferedWriter;
import java.io.File;
import java.io.FileOutputStream;
import java.io.FileWriter;
import java.io.InputStreamReader;
import java.net.URI;
import java.net.http.HttpClient;
import java.net.http.HttpRequest;
import java.net.http.HttpResponse;
import java.nio.charset.StandardCharsets;
import java.nio.file.Files;
import java.nio.file.Paths;
import java.security.MessageDigest;
import java.time.LocalDate;
import java.util.ArrayList;
import java.util.Collections;
import java.util.Comparator;
import java.util.HashMap;
import java.util.HashSet;
import java.util.List;
import java.util.Locale;
import java.util.Map;
import java.util.Set;
import java.util.zip.GZIPInputStream;

public class IngestBenchmark {
    private static final String MODE = "cpu"; // "cpu" (strict) or "io" (relaxed)
    private static final String DATA_DIR = "data";
    private static final String DOWNLOAD_DATE = "2025-01-01"; // set "" to disable downloads
    private static final int DOWNLOAD_HOURS = 1;
    private static final String OUTPUT_PARQUET_REL = "out/java_cpu_1h.parquet";
    private static final String OUTPUT_CHECKSUM_REL = "out/java_cpu_1h.sha256";
    private static final String[] INPUT_GLOBS = {};

    private static final ObjectMapper OBJECT_MAPPER = new ObjectMapper();

    private static class Stats {
        long totalLines;
        long validLines;
        long invalidLines;
    }

    private static class AggValue {
        long count;
        long sumRepoId;
    }

    private record Row(String eventDay, String eventType, long recordCount, long sumRepoId) {}

    public static void main(String[] args) throws Exception {
        String mode = MODE.toLowerCase(Locale.ROOT);
        if (!mode.equals("cpu") && !mode.equals("io")) {
            throw new IllegalArgumentException("MODE must be cpu or io");
        }
        boolean strict = mode.equals("cpu");

        String dataDir = resolveDataDir();
        String downloadDir = Paths.get(dataDir, "raw", "gharchive").toString();
        String outputParquet = Paths.get(dataDir, OUTPUT_PARQUET_REL).toString();
        String outputChecksum = Paths.get(dataDir, OUTPUT_CHECKSUM_REL).toString();

        List<String> downloaded = downloadGhArchive(DOWNLOAD_DATE, DOWNLOAD_HOURS, downloadDir);
        List<String> inputFiles = new ArrayList<>(resolveInputFiles(INPUT_GLOBS));
        inputFiles.addAll(downloaded);
        inputFiles = dedupeAndSort(inputFiles);
        if (inputFiles.isEmpty()) {
            throw new IllegalStateException("No input files found. Set INPUT_GLOBS and/or download settings.");
        }

        Stats stats = new Stats();
        Map<String, AggValue> aggregates = processFiles(inputFiles, strict, stats);
        List<Row> rows = sortedRows(aggregates);
        writeParquet(outputParquet, rows);
        String checksum = writeChecksum(outputChecksum, rows);

        System.out.println("[done]");
        System.out.println("mode=" + mode);
        System.out.println("input_files=" + inputFiles.size());
        System.out.println("total_lines=" + stats.totalLines);
        System.out.println("valid_lines=" + stats.validLines);
        System.out.println("invalid_lines=" + stats.invalidLines);
        System.out.println("aggregate_rows=" + rows.size());
        System.out.println("checksum=" + checksum);
        System.out.println("parquet=" + outputParquet);
        System.out.println("checksum_file=" + outputChecksum);
    }

    private static String resolveDataDir() {
        String[] candidates = {DATA_DIR, "data", "../data"};
        for (String c : candidates) {
            File f = new File(c);
            if (f.exists() && f.isDirectory()) {
                return c;
            }
        }
        return DATA_DIR;
    }

    private static void ensureParent(String filePath) throws Exception {
        File parent = new File(filePath).getParentFile();
        if (parent != null && !parent.exists()) {
            if (!parent.mkdirs()) {
                throw new RuntimeException("Failed to create dir: " + parent.getAbsolutePath());
            }
        }
    }

    private static List<String> downloadGhArchive(String dateStr, int hours, String downloadDir) throws Exception {
        List<String> out = new ArrayList<>();
        if (dateStr == null || dateStr.isBlank() || hours <= 0) {
            return out;
        }

        Files.createDirectories(Paths.get(downloadDir));
        HttpClient client = HttpClient.newBuilder().build();
        LocalDate date = LocalDate.parse(dateStr);

        for (int hour = 0; hour < hours; hour++) {
            LocalDate d = date.plusDays(hour / 24);
            int h = hour % 24;
            String datePart = d.toString();
            String fileName = datePart + "-" + h + ".json.gz";
            String url = "https://data.gharchive.org/" + fileName;
            String outPath = Paths.get(downloadDir, fileName).toString();

            if (Files.exists(Paths.get(outPath))) {
                System.out.println("[download] skip existing " + outPath);
                out.add(outPath);
                continue;
            }

            System.out.println("[download] " + url + " -> " + outPath);
            HttpRequest req = HttpRequest.newBuilder()
                    .uri(URI.create(url))
                    .header("User-Agent", "Mozilla/5.0")
                    .GET()
                    .build();
            HttpResponse<byte[]> resp = client.send(req, HttpResponse.BodyHandlers.ofByteArray());
            if (resp.statusCode() != 200) {
                throw new RuntimeException("Download failed with status " + resp.statusCode());
            }
            try (FileOutputStream fos = new FileOutputStream(outPath)) {
                fos.write(resp.body());
            }
            out.add(outPath);
        }
        return out;
    }

    private static List<String> resolveInputFiles(String[] globs) throws Exception {
        List<String> files = new ArrayList<>();
        for (String pattern : globs) {
            // Keep explicit globs simple for now; download paths cover primary use case.
            if (pattern != null && !pattern.isBlank()) {
                // No-op by design for now; fixed download set is used in this project baseline.
            }
        }
        Collections.sort(files);
        return files;
    }

    private static List<String> dedupeAndSort(List<String> files) {
        Set<String> seen = new HashSet<>(files);
        List<String> out = new ArrayList<>(seen);
        Collections.sort(out);
        return out;
    }

    private static Long getRepoId(JsonNode record) {
        JsonNode repo = record.get("repo");
        if (repo == null || repo.isNull()) return null;
        JsonNode id = repo.get("id");
        if (id == null || id.isNull()) return null;
        if (id.isIntegralNumber()) return id.asLong();
        if (id.isTextual()) {
            try {
                return Long.parseLong(id.asText());
            } catch (NumberFormatException ignored) {
                return null;
            }
        }
        return null;
    }

    private static boolean validateRecord(JsonNode record, boolean strict) {
        if (record == null || !record.isObject()) return false;
        String id = text(record, "id");
        String type = text(record, "type");
        String createdAt = text(record, "created_at");
        Long repoId = getRepoId(record);

        if (id.isEmpty() || type.isEmpty() || createdAt.isEmpty() || repoId == null) {
            return false;
        }
        if (strict) {
            if (!createdAt.contains("T") || !createdAt.endsWith("Z")) return false;
            if (type.length() > 100) return false;
        }
        return true;
    }

    private static String text(JsonNode n, String key) {
        JsonNode v = n.get(key);
        if (v == null || !v.isTextual()) return "";
        return v.asText();
    }

    private static Map<String, AggValue> processFiles(List<String> inputFiles, boolean strict, Stats stats) throws Exception {
        Map<String, AggValue> aggregates = new HashMap<>();
        for (String filePath : inputFiles) {
            System.out.println("[process] " + filePath);
            try (BufferedReader reader = new BufferedReader(new InputStreamReader(
                    new GZIPInputStream(Files.newInputStream(Paths.get(filePath))), StandardCharsets.UTF_8))) {
                String line;
                while ((line = reader.readLine()) != null) {
                    stats.totalLines++;
                    line = line.trim();
                    if (line.isEmpty()) {
                        stats.invalidLines++;
                        continue;
                    }
                    JsonNode record;
                    try {
                        record = OBJECT_MAPPER.readTree(line);
                    } catch (Exception e) {
                        stats.invalidLines++;
                        continue;
                    }
                    if (!validateRecord(record, strict)) {
                        stats.invalidLines++;
                        continue;
                    }
                    String createdAt = text(record, "created_at");
                    String eventDay = createdAt.substring(0, 10);
                    String eventType = text(record, "type").trim().toLowerCase(Locale.ROOT);
                    Long repoId = getRepoId(record);
                    if (repoId == null) {
                        stats.invalidLines++;
                        continue;
                    }

                    stats.validLines++;
                    String key = eventDay + "|" + eventType;
                    AggValue v = aggregates.getOrDefault(key, new AggValue());
                    v.count += 1;
                    v.sumRepoId += repoId;
                    aggregates.put(key, v);
                }
            }
        }
        return aggregates;
    }

    private static List<Row> sortedRows(Map<String, AggValue> aggregates) {
        List<Row> rows = new ArrayList<>();
        for (Map.Entry<String, AggValue> e : aggregates.entrySet()) {
            String[] parts = e.getKey().split("\\|", 2);
            String eventDay = parts[0];
            String eventType = parts.length > 1 ? parts[1] : "";
            rows.add(new Row(eventDay, eventType, e.getValue().count, e.getValue().sumRepoId));
        }
        rows.sort(Comparator.comparing(Row::eventDay).thenComparing(Row::eventType));
        return rows;
    }

    @SuppressWarnings("deprecation")
    private static void writeParquet(String outputParquet, List<Row> rows) throws Exception {
        ensureParent(outputParquet);
        Files.deleteIfExists(Paths.get(outputParquet));
        String schemaJson = """
                {
                  "type":"record",
                  "name":"aggregate_row",
                  "fields":[
                    {"name":"event_day","type":"string"},
                    {"name":"event_type","type":"string"},
                    {"name":"record_count","type":"long"},
                    {"name":"sum_repo_id","type":"long"}
                  ]
                }
                """;
        Schema avroSchema = new Schema.Parser().parse(schemaJson);
        Configuration conf = new Configuration();

        try (ParquetWriter<GenericRecord> writer = AvroParquetWriter.<GenericRecord>builder(new Path(outputParquet))
                .withSchema(avroSchema)
                .withCompressionCodec(CompressionCodecName.ZSTD)
                .withConf(conf)
                .build()) {
            for (Row row : rows) {
                GenericRecord rec = new GenericData.Record(avroSchema);
                rec.put("event_day", row.eventDay());
                rec.put("event_type", row.eventType());
                rec.put("record_count", row.recordCount());
                rec.put("sum_repo_id", row.sumRepoId());
                writer.write(rec);
            }
        }
    }

    private static String writeChecksum(String outputChecksum, List<Row> rows) throws Exception {
        ensureParent(outputChecksum);
        Files.deleteIfExists(Paths.get(outputChecksum));
        MessageDigest digest = MessageDigest.getInstance("SHA-256");
        for (Row row : rows) {
            String line = row.eventDay() + "|" + row.eventType() + "|" + row.recordCount() + "|" + row.sumRepoId() + "\n";
            digest.update(line.getBytes(StandardCharsets.UTF_8));
        }
        StringBuilder sb = new StringBuilder();
        for (byte b : digest.digest()) {
            sb.append(String.format("%02x", b));
        }
        String checksum = sb.toString();
        try (BufferedWriter w = new BufferedWriter(new FileWriter(outputChecksum, StandardCharsets.UTF_8))) {
            w.write(checksum);
            w.write("\n");
        }
        return checksum;
    }
}
